init

bendanzhentan opened the pull request.

bendanzhentan commented on cmd/bot/main.go:
```diff
@@ -0,0 +1,368 @@
+			continue
+		}
+
+		if latestNumber < uint64(cfg.Misc.ConfirmBlocks) {
```
`cfg.Misc.ConfirmBlocks` specifies how many blocks behind the latest block, similar to op-node's [`sequencer.l1-confs`](https://github.com/ethereum-optimism/optimism/blob/develop/op-node/flags/flags.go#L142-L148).

This is to avoid issues with recent blocks that might become reorganized (reorged)

It’s like keeping a safe distance from the latest blocks to ensure stability and accuracy.

bendanzhentan commented on cmd/bot/main.go:
```diff
@@ -0,0 +1,368 @@
+		}
+
+		log.Info("Fetching logs from blocks", "fromBlock", fromBlockNumber, "toBlock", toBlockNumber)
+		logs, err := getLogs(client, fromBlockNumber, toBlockNumber, common.HexToAddress(cfg.Misc.L2StandardBridgeBot))
```
```go
		toBlockNumber := new(big.Int).Add(fromBlockNumber, big.NewInt(cfg.Misc.LogFilterBlockRange))
```

`toBlockNumber` should be `fromBlockNumber + cfg.Misc.LogFilterBlockRange`

bendanzhentan commented on cmd/bot/main.go:
```diff
@@ -0,0 +1,368 @@
+		}
+
+		l2ScannedBlock.Number = toBlockNumber.Int64()
+		result := db.Where("number >= 0").Updates(l2ScannedBlock)
```
I agree with you. But when I use `Save` method, the program reports a warning says that WHERE clause should be set. This workaround has worked for me. Let's troubleshoot this issue ln the future.

bendanzhentan commented on cmd/bot/main.go:
```diff
@@ -0,0 +1,368 @@
+		}
+
+		unprovens := make([]core.L2ContractEvent, 0)
+		if result := db.Order("id asc").Where("proven = false").Limit(limit).Find(&unprovens); result.Error != nil {
```
Hi @welkin22, I'm not quite sure how this type of issue will occur.

1. L2ContractEvents are stored in scanning order, which means they are ordered by on-chain order.

2. When querying `L2ContractEvent`, we sort them by the auto-increment `id`, so they are returned in the on-chain order.

3. When a withdrawal cannot be proven or finalized, then subsequent withdrawals are also  unable to prove or finalize.

welkin22 replied:
The 3 points you mentioned here can exactly prove the cause of the problem. Let me give you an example:
1. Our program has already processed up to id 12000.
2. At this point, someone maliciously attacks and creates 1000 unprovable transactions (although I don't know how this attack can happen, but we assume it is possible), so data from id 12001 to 13001 will not be processed.
3. Our program will keep trying to process data between 12001 and 13001, but with no progress.
4. Transactions initiated by normal users after 13001 will not be processed.

bendanzhentan commented on core/types.go:
```diff
@@ -0,0 +1,17 @@
+
+type L2ContractEvent struct {
+	ID              uint   `gorm:"primarykey"`
+	BlockTime       int64  `gorm:"type:integer;not null;index"`
```
Thanks, will add soon.

bendanzhentan replied:
Updated at [02dc445](https://github.com/bnb-chain/opbnb-bridge-bot/pull/1/commits/02dc4450b7536eedb5e8107bc15cc253224c09ce), PTAL

welkin22 replied:
Naming fields is great, but what I actually meant was to give indexes a name. You can refer to: https://gorm.io/docs/indexes.html
For example:
```
Name2 string `gorm:"index:idx_name,unique"`
```

bendanzhentan commented on cmd/bot/main.go:
```diff
@@ -0,0 +1,368 @@
+	if err != nil {
+		return fmt.Errorf("failed to migrate l2_contract_events: %w", err)
+	}
+	db.Exec("ALTER TABLE l2_contract_events ADD UNIQUE (block_hash, log_index)")
```
Could you please help me to rewrite it using gorm ways? Since the constraint "l2_contract_events_block_hash_log_index_key" combines block_hash and log_index, I don't know how to define it in native gorm ways.

bendanzhentan committed:
init

bendanzhentan committed:
feat: use gorm native ways

bendanzhentan committed:
chore: reduce ticker duration of processor

bendanzhentan committed:
docs: update README.md

welkin22 reviewed (approved):
LGTM

bendanzhentan committed:
contracts: remove delegationFee view function

bendanzhentan committed:
contracts: add more fields for WithdrawTo event

bendanzhentan committed:
contracts: fix bep20 withdrawal

bendanzhentan committed:
contracts: define withdrawFee() with onlyOwner

bendanzhentan committed:
contracts: emit Bot events as last one

bendanzhentan committed:
chore: split code into run.go

owen-reorg commented on contracts/src/L2StandardBridgeBot.sol:
```diff
@@ -0,0 +1,72 @@
+
+    // withdrawFee withdraw the delegation fee vault to _recipient address, only owner can call this function.
+    function withdrawFee(address _recipient) external onlyOwner {
+        payable(_recipient).transfer(address(this).balance);
```
plz use `call` to send ether. 
ref: https://solidity-by-example.org/sending-ether/

owen-reorg commented:
the app can't compile, plz set at least a build ci

owen-reorg commented on core/types.go:
```diff
@@ -0,0 +1,17 @@
+	ID              uint   `gorm:"primarykey"`
+	BlockTime       int64  `gorm:"type:integer;not null;index:idx_l2_contract_events_block_time"`
+	BlockHash       string `gorm:"type:varchar(256);not null;uniqueIndex:idx_l2_contract_events_block_hash_log_index_key,priority:1;"`
+	ContractAddress string `gorm:"type:varchar(256);not nul"`
```
```suggestion
	ContractAddress string `gorm:"type:varchar(256);not null"`
```

owen-reorg commented on contracts/src/L2StandardBridgeBot.sol:
```diff
@@ -0,0 +1,72 @@
+        uint256 _amount,
+        uint32 _minGasLimit,
+        bytes calldata _extraData
+    ) public payable {
```
The user may compose malicious tx for the bot runner to finalize, and these limitations can't be hardcoded in the contracts. So we should set filter for events to handle:
  - `_l2Token` in known list
  - `_minGasLimit` can't be bigger than known value. In practice, we'll set a fixed value in the front end
  - `_extraData` should be empty

txns from bridge UI will match the filter while custom txn sent directly by users may not.

bendanzhentan replied:
Hi @owen-reorg I have created another PR https://github.com/bnb-chain/opbnb-bridge-bot/pull/8

With https://github.com/bnb-chain/opbnb-bridge-bot/pull/8, I think it can handle malicious cases alraedy.

Then I think this suggesting filter may be not necessary for a malicious withdrawal, since the contract will return a revert error when proving or finalizing. Afterward, we will record the revert-error and [skip the reverted withdrawals](https://github.com/bnb-chain/opbnb-bridge-bot/pull/8/files#diff-279bc857f9ebbe8ad5b753f321c54f1e4ce671d6f489831783e71cbf46e3c69cR97) later.

owen-reorg commented on core/types.go:
```diff
@@ -0,0 +1,17 @@
+	LogIndex        int    `gorm:"type:integer;not null;uniqueIndex:idx_l2_contract_events_block_hash_log_index_key,priority:2"`
+	EventSignature  string `gorm:"type:varchar(256);not null"`
+	Proven          bool   `gorm:"type:boolean;not null;default:false"`
+	Finalized       bool   `gorm:"type:boolean;not null;default:false"`
```
If a withdrawal txn is not expected, we may need to ignore it. I suggest to add a field for this.

bendanzhentan replied:
Another PR has been created to apply this suggestion https://github.com/bnb-chain/opbnb-bridge-bot/pull/8

owen-reorg commented on bot.toml:
```diff
@@ -0,0 +1,25 @@
+name = "db_name"
+
+[l1-contracts]
+address-manager                 = "0x1111111111111111111111111111111111111111"
```
It's determined for both testnet and mainnet.

1. Remove unnecessary configurations.
2. List both testnet and mainnet configurations here. Use testnet as the default and comment out the mainnet configuration.

owen-reorg commented on core/processor.go:
```diff
@@ -0,0 +1,544 @@
+		return fmt.Errorf("hashMesaageHash err: %v", err)
+	}
+
+	l2OutputIndex, outputProposal, err := b.getL2OutputAfter(l2BlockNumber)
```
I believe we can use the latest L2 block number in L2OutputOracle. Can you please give it a try? If we use an earlier block, the performance of getProof will be worse. Additionally, nodes running in certain modes may remove the early trie data, making it impossible to obtain the proof.

owen-reorg commented on core/processor.go:
```diff
@@ -0,0 +1,544 @@
+	}
+
+	// Retrieve the private key from the environment variable OPBNB_BRIDGE_BOT_PRIVKEY
+	botPrivkey, err := crypto.HexToECDSA(os.Getenv("OPBNB_BRIDGE_BOT_PRIVKEY"))
```
Since you already use `os.ExpandEnv` in `config.go`, I suggest making the bot_privatekey a configuration option in a `toml` file. Users can choose to write the private key directly in the config or use an environment variable instead of a fixed value.

owen-reorg commented on core/processor.go:
```diff
@@ -0,0 +1,544 @@
+	pubKeyHash := crypto.Keccak256(pubKeyBytes[1:])[12:]
+	fromAddress := common.HexToAddress(hexutil.Encode(pubKeyHash))
+
+	const gasPrice = 9000000000 // 9 GWei
```
We should reconsider this; 9 gwei is too high for practical use.

owen-reorg commented on core/processor.go:
```diff
@@ -0,0 +1,544 @@
+		b.L1Contracts.OptimismPortalProxy,
+		b.L1Client,
+	)
+	signedTx, err := optimismPortalTransactor.ProveWithdrawalTransaction(
```
will it dry run before send?
what if the tx fails?
what if the tx pending for a long time?

bendanzhentan replied:
1. It will estimate gas before sending. Eth_estimateGas can be considered a dry run.

2. No retry for tx failures. These unproven withdrawals will be rescanned by the processor goroutine and retried to prove.

3. As (2) says, I think failed transactions will be retried by the processor goroutine.

I am finding a way to enhance the transaction management, please touch me if you have some ideas.

owen-reorg commented on core/processor.go:
```diff
@@ -0,0 +1,544 @@
+}
+
+// FinalizeMessage https://github.com/ethereum-optimism/optimism/blob/d90e7818de894f0bc93ae7b449b9049416bda370/packages/sdk/src/cross-chain-messenger.ts#L1611
+func (b *Processor) FinalizeMessage(botDelegatedWithdrawToEvent *L2ContractEvent) error {
```
too many duplicate codes, should use a function here.

owen-reorg commented:
There is a lack of main loop logic. Did you forget to submit any files?

bendanzhentan committed:
apply review suggestions

1. contracts: transfer fee via call
2. cmd: git add run.go
3. core: add SignerConfig for signer and gas price
4. core: reduce duplicated code in ProveWithdrawalTransaction and FinalizeWithdrawalTransaction
5. core: fix type definition ContractAddress
6. ci: add workflow build.yaml

bendanzhentan commented:
> There is a lack of main loop logic. Did you forget to submit any files?

Yes, the fun fact is that `.gitignore` includes `cmd/` so that I didn't realize that `cmd/bot/run.go` didn't include in git.

I make some updates at [ddfaf66](https://github.com/bnb-chain/opbnb-bridge-bot/pull/1/commits/ddfaf662fda5410c4e53f0af753d161b6b465ca9) according to the below review suggestions:

1. contracts: transfer fee via call https://github.com/bnb-chain/opbnb-bridge-bot/pull/1#discussion_r1399183436
2. cmd: git add run.go https://github.com/bnb-chain/opbnb-bridge-bot/pull/1#issuecomment-1819090420
3. core: add SignerConfig for signer and gas price https://github.com/bnb-chain/opbnb-bridge-bot/pull/1#discussion_r1399276930
4. core: reduce duplicated code in ProveWithdrawalTransaction and FinalizeWithdrawalTransaction https://github.com/bnb-chain/opbnb-bridge-bot/pull/1#discussion_r1399286718
5. core: fix type definition ContractAddress https://github.com/bnb-chain/opbnb-bridge-bot/pull/1#discussion_r1399233944
6. ci: add workflow build.yaml https://github.com/bnb-chain/opbnb-bridge-bot/pull/1#issuecomment-1819090420

owen-reorg commented on cmd/bot/run.go:
```diff
@@ -0,0 +1,305 @@
+	result := db.Order("number desc").Last(&l2ScannedBlock)
+	if result.Error != nil {
+		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
+			db.Create(&l2ScannedBlock)
```
shall we set a `start number` in config instead of start from block #0?

owen-reorg commented on cmd/bot/run.go:
```diff
@@ -0,0 +1,305 @@
+			continue
+		} else if len(unprovens) > 0 {
+			for _, unproven := range unprovens {
+				if unproven.BlockTime+cfg.Misc.ProposeTimeWindow < time.Now().Unix() {
```
Normally too deep nested code is hard to read. I suggest rewrite it by handle earlier with `esle` condition.

E.g. 

```
if unproven.BlockTime+cfg.Misc.ProposeTimeWindow >= time.Now().Unix() {
    continue
}
original_if_code_block()
```

bendanzhentan replied:
Another PR has been created to refactor this function, to make it more readable. PTAL https://github.com/bnb-chain/opbnb-bridge-bot/pull/7

owen-reorg committed:
Update README.md

owen-reorg committed:
fix: withdraw bep20 error

owen-reorg committed:
add expect event

owen-reorg committed:
fix ci

bendanzhentan committed:
docs: add motivation to README.md

bendanzhentan committed:
Merge pull request #2 from bnb-chain/bep20-test

fix: withdraw bep20 error
